src/Facebook/FileUpload/FacebookResumableUploader.php

```
@@ -0,0 +1,208 @@
+ * Class FacebookResumableUploader
+ *
+ * @package Facebook
+ */
+class FacebookResumableUploader
+{
+    /**
+     * @var FacebookApp
+     */
+    protected $app;
+
+    /**
+     * @var string
+     */
+    protected $accessToken;
+
+    /**
+     * @var FacebookClient The Facebook client service.
+     */
+    protected $client;
+
+    /**
+     * @var int Max retry times
+     */
+    protected $maxTransferTries;
+
+    /**
+     * @param FacebookApp $app
+     * @param string $accessToken
+     * @param int $maxTransferTries
+     */
+    public function __construct(FacebookApp $app, $accessToken, $maxTransferTries = 5)
+    {
+        $this->accessToken = $accessToken;
+        $this->app = $app;
+        $this->client = new FacebookClient();
+        $this->maxTransferTries = $maxTransferTries;
+    }
+
+    /**
+     * Upload - phase start
+     *
+     * @param $endpoint
+     * @param $filePath
+     * @return FacebookTransferChunk
+     *
+     * @throws FacebookUploadException
+     * @throws FacebookSDKException
+     */
+    public function start($endpoint, $filePath)
+    {
+        $fileSize = filesize($filePath);
+
+        $startReqParams = [
+            'upload_phase' => 'start',
+            'file_size' => $fileSize,
+        ];
+
+        $request = new FacebookRequest(
+            $this->app,
+            $this->accessToken,
+            'POST',
+            $endpoint,
+            $startReqParams
+        );
+
+        try {
+            $res= $this->client->sendRequest($request)->getDecodedBody();
+
+            $firstTransferChunk = new FacebookTransferChunk(
+                $filePath,
+                $res['upload_session_id'],
+                $res['video_id'],
+                $res['start_offset'],
+                $res['end_offset'] - $res['start_offset']
+            );
+
+            return $firstTransferChunk;
+        } catch (FacebookResponseException $e) {
+            throw new FacebookUploadException($e->getResponse());
```

Sorry - this comment applies to the `transfer` phase. Start phase should let the exception bubble up.